src/mod/applications/mod_http_cache/prometheus_metrics.cpp

```diff
@@ -0,0 +1,157 @@
+#include <map>
+#include <string>
+#include "prometheus_metrics.h"
+
+
+class prometheus_metrics
+{
+public:
+	class auto_lock
+	{
+	public:
+		auto_lock(switch_mutex_t* mutex) : _mutex(mutex)
+		{
+			switch_mutex_lock(_mutex);
+		}
+		~auto_lock()
+		{
+			switch_mutex_unlock(_mutex);
+		}
+		switch_mutex_t* _mutex;
+	};
+	
+	prometheus_metrics(switch_loadable_module_interface_t **module_interface, switch_memory_pool_t* pool) :
+		_pool(pool),
+		_module_interface(module_interface),
+		_download_fail_count(0),
+		_download_bucket_0_300(0),
+		_download_bucket_300_500(0),
+		_download_bucket_500_800(0),
+		_download_bucket_800_1300(0),
+		_download_bucket_1300_2100(0),
+		_download_bucket_2100_3400(0),
+		_download_bucket_3400_5500(0),
+		_download_bucket_5500_inf(0),
+		_download_bucket_sum(0),
+		_download_bucket_count(0)
+	{
+		switch_mutex_init(&_mutex, SWITCH_MUTEX_NESTED, _pool);
+	}
+	
+	~prometheus_metrics()
+	{
+		switch_mutex_destroy(_mutex);
+	}
+
+	void increment_download_duration(unsigned int duration)
+	{
+		auto_lock lock(_mutex);
+
+		_download_bucket_sum += duration;
+		_download_bucket_count++;
+
+		if (duration < 300) {
+			_download_bucket_0_300++;
+		} else if (300 <= duration && duration < 500) {
+			_download_bucket_300_500++;
+		} else if (500 <= duration && duration < 800) {
+			_download_bucket_500_800++;
+		} else if (800 <= duration && duration < 1300) {
+			_download_bucket_800_1300++;
+		} else if (1300 <= duration && duration < 2100) {
+			_download_bucket_1300_2100++;
+		} else if (2100 <= duration && duration < 3400) {
+			_download_bucket_2100_3400++;
+		} else if (3400 <= duration && duration < 5500) {
+			_download_bucket_3400_5500++;
+		} else {
+			_download_bucket_5500_inf++;
```

Comment:
Just a note here and not a blocker. Maybe we can improve the implementation here. We could either use a combination of std::map and use lower_bound and upper_bound to get the range plus use std::accumulate to compute the bucket. Another is to use a range base key class for std::map then store those value.

Reply:
Agree. :)
This can be better. I've fixed it.